File: backend/src/main/java/com/back/back9/domain/common/vo/money/Money.java
package com.back.back9.domain.common.vo.money;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

public class Money {

    private static final int SCALE = 8; // 소수점 8자리까지 표현
    private static final RoundingMode ROUNDING_MODE = RoundingMode.HALF_EVEN;

    private final BigDecimal amount;

    private Money(BigDecimal amount) {
        if (amount == null) {
            throw new IllegalArgumentException("Amount cannot be null");
        }
        this.amount = amount.setScale(SCALE, ROUNDING_MODE);
    }

    public static Money of(BigDecimal amount) {
        return new Money(amount);
    }

    public static Money of(long value) {
        return new Money(BigDecimal.valueOf(value));
    }

    public static Money zero(){
        return new Money(BigDecimal.ZERO);
    }

    public BigDecimal toBigDecimal() {
        return amount;
    }

    public Money add(Money other) {
        return new Money(this.amount.add(other.amount));
    }

    public Money subtract(Money other) {
        return new Money(this.amount.subtract(other.amount));
    }

    public Money multiply(BigDecimal multiplier) {
        return new Money(this.amount.multiply(multiplier));
    }

    public Money divide(BigDecimal divisor) {
        if (divisor.compareTo(BigDecimal.ZERO) == 0) {
            return Money.zero();
        }
        return new Money(this.amount.divide(divisor, SCALE, ROUNDING_MODE));
    }

    public boolean isGreaterThanZero() {
        return this.amount.compareTo(BigDecimal.ZERO) > 0;
    }

    public boolean isGreaterThanOrEqual(Money other) {
        return this.amount.compareTo(other.amount) >= 0;
    }

    @Override
    public String toString() {
        return amount.toPlainString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Money money)) return false;
        return amount.compareTo(money.amount) == 0; // 금액이 같으면 동일
    }

    @Override
    public int hashCode() {
        return Objects.hash(amount); // 금액 기준으로 해시값 생성
    }

    public boolean isZero() {
        return this.amount.compareTo(BigDecimal.ZERO) == 0;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    public Money divide(Money amount, int i, RoundingMode roundingMode) {
        return new Money(this.amount.divide(amount.amount, i, roundingMode));
    }
}

File: backend/src/main/kotlin/com/back/back9/domain/common/vo/money/Money.kt
package com.back.back9.domain.common.vo.money

import java.math.BigDecimal
import java.math.RoundingMode

@JvmInline
value class Money private constructor(val amount: BigDecimal) {

    init {
        require(amount >= BigDecimal.ZERO) { "금액은 음수가 될 수 없습니다." }
    }

    fun toBigDecimal(): BigDecimal = amount

    fun add(other: Money): Money = Money(amount.add(other.amount))

    fun subtract(other: Money): Money = Money(amount.subtract(other.amount))

    fun multiply(multiplier: BigDecimal): Money = Money(amount.multiply(multiplier))

    fun divide(divisor: BigDecimal): Money =
        if (divisor == BigDecimal.ZERO) zero()
        else Money(amount.divide(divisor, SCALE, ROUNDING_MODE))

    val isPositive: Boolean get() = amount > BigDecimal.ZERO
    fun isGreaterThanZero(): Boolean = amount > BigDecimal.ZERO

    fun isGreaterThanOrEqual(other: Money): Boolean = amount >= other.amount

    val isZero: Boolean get() = amount.compareTo(BigDecimal.ZERO) == 0

    override fun toString(): String = amount.toPlainString()

    companion object {
        private const val SCALE = 8
        private val ROUNDING_MODE = RoundingMode.HALF_EVEN

        fun of(amount: BigDecimal): Money = Money(amount.setScale(SCALE, ROUNDING_MODE))

        fun of(value: Long): Money = Money(BigDecimal.valueOf(value).setScale(SCALE, ROUNDING_MODE))

        fun zero(): Money = Money(BigDecimal.ZERO.setScale(SCALE, ROUNDING_MODE))
    }
}
